### PR TITLE
feat(SourceMetrics): kg source metrics custom function

### PR DESCRIPTION
**Pull request by bohandley**
### ✨ Description

**Related issue(s):** https://github.com/grafana/metrics-drilldown/issues/1131
Addresses #1131 (see "Open question for KG" below; final closure depends on KG's choice).

[KG companion PR](https://github.com/grafana/asserts-app-plugin/pull/3278)

Allow Knowledge Graph (KG) to override the default PromQL aggregation function per source metric via the `<SourceMetrics />` exposed component. For a gauge insight like `PrometheusRemoteWriteDesiredShards`, KG can pass `customFunction: 'max_over_time'` and GMD will render the source metric with that function instead of the default `avg(...)`. The override flows from KG via the React prop and survives the embedded-to-standalone handoff through multi-value URL sync. The hint is opt-in; entries without `customFunction` retain the existing type-driven default.

#### Consensus arrived: we will keep the multiple series. KG will add more filters to reduce the series.

<table>
  <tr>
    <th align="center">MetricsReducer</th>
    <th align="center">MetricScene</th>
  </tr>
  <tr>
    <td>Screenshot 2026-06-12 at 4 35 40 PM</td>
    <td>Screenshot 2026-06-12 at 4 35 45 PM</td>
  </tr>
</table>

Range-vector functions (`max_over_time`, `min_over_time`) operate per series. The current implementation emits the function literally:

```
max_over_time(prometheus_remote_storage_shards_desired{labels}[$__rate_interval])
```

When the metric has multiple underlying label combinations (eg `remote_name`, `url`), the panel shows one line per series, all sharing the same legend. The default `avg(...)` collapses these into a single line because it aggregates across series. See attached screenshot for the multi-series render against `prometheus_remote_storage_shards_desired` on the alloy / hosted-exporters entity.

KG needs to decide:

1. Keep the bare `max_over_time(...)` as shipped here (per-series view).
2. Wrap with an outer instant aggregation, eg `avg(max_over_time(metric[interval]))`, to preserve the single-line look of `avg`.
3. Whether the choice should differ between MetricsReducer (the multi-tile list) and MetricScene (the single-metric detail).

This PR ships option 1. The outer-wrap follow-up is gated on KG's answer.

### 📖 Summary of the changes

- Adds opt-in `customFunction?: string` per source-metric entry. Threads through `<SourceMetrics />` into `DataTrail.state.sourceMetrics`, through `GmdVizPanel.queryConfig`, into the timeseries and stat query builders. Precedence in the builders: URL `customFunction` wins over the configurator's per-metric pref, which wins over the type-driven default (avg/sum/count). Range-vector functions (`max_over_time`, `min_over_time`) use `customRateInterval` for the interval and fall back to the type default in group-by mode because `fn(metric[interval]) by (label)` is not legal PromQL.
- URL sync extended: `customFunction` is a multi-value query param with value format `{fn}-{metricName}`, so multi-metric insights round-trip across the embedded ↔ standalone boundary. The single-value `customRateInterval` URL shape shipped in #1130 is untouched; a follow-up will align them once #1058 lands.
- Unit tests added for the timeseries and stat query builders covering range-function emission, precedence, and fall-through.

### 🧪 How to test?

1. Pair with the asserts-app-plugin branch that extends `AssertionSourceMetric`: https://github.com/grafana/asserts-app-plugin/pull/3278.

Add a temporary smoke patch in `asserts-app-plugin/src/features/EntityDetails/hooks/useEntityDetailsData.ts`, after the `useAssertionSourceMetrics` call. Make sure `useMemo` is imported from `react` (it already is in this file).

```typescript
const sourceMetricsWithCustomFunction = useMemo(
  () =>
    sourceMetrics?.map((s) =>
      s.metricName === 'prometheus_remote_storage_shards_desired'
        ? { ...s, customFunction: 'max_over_time' }
        : s
    ),
  [sourceMetrics]
);
```

Return `sourceMetricsWithCustomFunction` in place of the raw `sourceMetrics` from the hook.

2. In RCA Workbench, open the `PrometheusRemoteWriteDesiredShards` alert for an alloy entity (eg `alloy` in `hosted-exporters`).
3. Inspect the rendered query (panel menu, Inspect, Query) for `prometheus_remote_storage_shards_desired`. Confirm `max_over_time(prometheus_remote_storage_shards_desired{labels}[$__rate_interval])`.
4. Check the host URL contains `gmd-customFunction=max_over_time-prometheus_remote_storage_shards_desired`.
5. Click the `Metrics Drilldown` button. Standalone URL contains `customFunction=max_over_time-prometheus_remote_storage_shards_desired` (no `gmd-` prefix). Standalone panel renders the same query.
6. Navigate to a different metric in MetricsReducer and confirm the URL drops the override when no entry matches, then restores it on back-navigation.
7. `pnpm test` for the two updated query-builder suites.

### CODE REVIEW

File: src/shared/GmdVizPanel/types/timeseries/getTimeseriesQueryRunnerParams.ts [hunk 1]
```diff
@@ -87,7 +134,8 @@ function buildQueriesWithPresetFunctions({
       logger.warn(`[getTimeseriesQueryRunnerParams] Unknown PromQL function "${fn}", skipping query.`);
       continue;
     }
-    const query = entry.fn({ expr });
+    const isRangeFn = entry.name === 'max_over_time' || entry.name === 'min_over_time';
+    const query = isRangeFn ? entry.fn({ expr, interval }) : entry.fn({ expr });
     const fnName = metric.type === 'counter' ? `${entry.name}(rate)` : entry.name;
```
[the remark this answers is not in the document]
Comment (bohandley) [reply src/shared/GmdVizPanel/types/timeseries/getTimeseriesQueryRunnerParams.ts]: We are trusting KG to send correct function and type. If the incorrect match is sent, then we do not want to swallow the error.

File: src/shared/GmdVizPanel/types/stat/getStatQueryRunnerParams.ts [hunk 2]
```diff
@@ -52,7 +81,8 @@ function buildQueriesWithPresetFunctions({
       logger.warn(`[getStatQueryRunnerParams] Unknown PromQL function "${fn}", skipping query.`);
       continue;
     }
-    const query = entry.fn({ expr });
+    const isRangeFn = entry.name === 'max_over_time' || entry.name === 'min_over_time';
+    const query = isRangeFn ? entry.fn({ expr, interval }) : entry.fn({ expr });
     const fnName = isRateQuery ? `${entry.name}(rate)` : entry.name;
```
[the remark this answers is not in the document]
Comment (bohandley) [reply src/shared/GmdVizPanel/types/stat/getStatQueryRunnerParams.ts]: We are trusting KG to send correct function and type. If the incorrect match is sent, then we do not want to swallow the error.

File: src/shared/GmdVizPanel/types/stat/__tests__/getStatQueryRunnerParams.test.ts [hunk 3]
```diff
@@ -89,4 +89,98 @@ describe('getStatQueryRunnerParams(options)', () => {
...
+
+    expect(result.queries).toStrictEqual([
+      {
+        refId: 'http_requests_total-avg(rate)',
+        expr: 'avg(rate(http_requests_total{__ignore_usage__="", ${filters:raw}}[$__rate_interval]))',
+        legendFormat: 'avg(rate)',
+        fromExploreMetrics: true,
+      },
+    ]);
+  });
```
[the remark this answers is not in the document]
Comment (bohandley) [reply src/shared/GmdVizPanel/types/stat/__tests__/getStatQueryRunnerParams.test.ts:153]: We are trusting KG to send correct function and type. If the incorrect match is sent, then we do not want to swallow the error.

File: src/shared/GmdVizPanel/types/timeseries/__tests__/getTimeseriesQueryRunnerParams.test.ts [hunk 4]
```diff
@@ -101,6 +101,97 @@ describe('getTimeseriesQueryRunnerParams(options)', () => {
...
+
+      expect(result.queries).toStrictEqual([
+        {
+          refId: 'http_requests_total-avg(rate)',
+          expr: 'avg(rate(http_requests_total{__ignore_usage__="", ${filters:raw}}[$__rate_interval]))',
+          legendFormat: 'avg(rate)',
+          fromExploreMetrics: true,
+        },
+      ]);
+    });
```
[the remark this answers is not in the document]
Comment (bohandley) [reply src/shared/GmdVizPanel/types/timeseries/__tests__/getTimeseriesQueryRunnerParams.test.ts:162]: We are trusting KG to send correct function and type. If the incorrect match is sent, then we do not want to swallow the error.

File: src/shared/GmdVizPanel/types/timeseries/getTimeseriesQueryRunnerParams.ts [hunk 5]
```diff
@@ -50,18 +50,61 @@ function buildGroupByQueries({
     fn = 'count';
   }
 
+  // KG-supplied customFunction (issue #1131) wins over the type default for instant aggregations.
+  // The group-by path calls `promql[fn]({expr, by})` directly on tsqtsq, so the override is only
+  // honored for instant aggregations that tsqtsq exposes and that accept a `by` clause. Range-vector
+  // functions and custom-shaped entries (histogram_quantile, time-*) fall back to the type default
+  // in group-by mode.
```
[the remark this answers is not in the document]
Comment (bohandley) [reply src/shared/GmdVizPanel/types/timeseries/getTimeseriesQueryRunnerParams.ts]: Comment is updated.

File: src/AppDataTrail/sourceMetricsUrlSync.ts [hunk 6]
```diff
@@ -0,0 +1,64 @@
+import { type SceneObjectUrlValues } from '@grafana/scenes';
+
+import { type SourceMetrics } from '../exposedComponents/SourceMetrics/types';
+
+// Multi-value URL value parser for the customFunction key.
```
Comment (bohandley) [inline src/AppDataTrail/sourceMetricsUrlSync.ts:5]: Here's the multi metric implementation for function. We need this for type as well.

File: src/shared/GmdVizPanel/types/timeseries/getTimeseriesQueryRunnerParams.ts [hunk 7]
```diff
@@ -78,25 +98,22 @@ function buildQueriesWithPresetFunctions({
...
     defaultPromqlFn = 'count';
   }
+
+  const interval = queryConfig.customRateInterval ?? '$__rate_interval';
+  const isCounter = metric.type === 'counter';
+  const customFn = queryConfig.customFunction;
+
+  if (customFn) {
+    return [buildCustomFunctionQuery(metric.name, customFn, expr, interval, isCounter)];
+  }
```
Comment (shelldandy) [inline src/shared/GmdVizPanel/types/timeseries/getTimeseriesQueryRunnerParams.ts:108]: Counter + range-vector customFunction produces invalid PromQL.

**Review by shelldandy — CHANGES_REQUESTED** [review]
## Summary

Well-structured feature that threads an opt-in customFunction override from KG through URL sync, state, and into both query builders. The URL serialization, extracted helpers, and test coverage for the gauge path are solid. However, there is an ungated edge case where a range-vector customFunction (e.g. max_over_time) on a counter metric produces invalid PromQL because the expr is already rate()-wrapped before the range function adds [interval].

## Issues

- [warning] src/shared/GmdVizPanel/types/timeseries/getTimeseriesQueryRunnerParams.ts:106: **Counter + range-vector customFunction produces invalid PromQL.**

For counters, `expr` is already `rate(metric{...}[$__rate_interval])` (line 26). When `customFn` is a range-vector function like `max_over_time`, `buildCustomFunctionQuery` emits:

```
max_over_time(rate(metric{...}[$__rate_interval])[$__rate_interval])
```

A range selector `[interval]` cannot follow `rate()` (which yields an instant vector) — this is a PromQL parse error. The same issue applies in `buildGroupByQueries` (line 64) and in the stat builder.

Currently the documented use case is gauge-only, but nothing gates `customFunction` by metric type — KG passes any string and it flows straight into the query. A future misconfiguration or expanded use would silently break.

Consider either:
1. Skipping the `rate()` pre-wrap when `customFn` is a range-vector function (use the raw `expression` instead of `expr`), or
2. Falling back to the type-default instant function with a `logger.warn`, mirroring the fallback you already chose for the `by` clause case.
- [warning] src/shared/GmdVizPanel/types/timeseries/__tests__/getTimeseriesQueryRunnerParams.test.ts:143: **Missing test: counter + range-vector customFunction.**

The tests cover gauge+range (`max_over_time`) and counter+instant (`avg`), but there's no test for counter+range (e.g. counter with `customFunction: 'max_over_time'`). That's the exact path that produces invalid PromQL. Adding this test would surface the bug and document the intended behavior for that combination. Same gap exists in the stat test suite.

File: src/shared/GmdVizPanel/types/timeseries/getTimeseriesQueryRunnerParams.ts (hunk 7 above)
Comment (bohandley) [reply src/shared/GmdVizPanel/types/timeseries/getTimeseriesQueryRunnerParams.ts:108]: Good catch!  We are ok with allowing KG to pass a function that may create invalid PromQL if the metric type does not support it. If KG passes a function that does not fit the metric, we want that to surface as a real Prometheus error on the panel rather than be masked. [The documented case is for gauges here in no. 3](https://docs.google.com/document/d/1CkvtZGRoJBFOhb6Cd0jtUlIy2-E4pYXNtSb5N20GJww/edit?tab=t.0).

Comment (bohandley) [reply src/shared/GmdVizPanel/types/timeseries/getTimeseriesQueryRunnerParams.ts:108]: Adding tests to show documentation of this 👍🏻

**Review by shelldandy — APPROVED** [review]
after discussing earlier seems like the potentially bad promql queries are good since they'd fall under the KG team to ensure those overrides comes properly to us
